packages/flutter/lib/src/cupertino/magnifier.dart

```diff
@@ -0,0 +1,308 @@
+    this.dragResistance = 10.0,
+    this.hideBelowThreshold = -48.0,
+    this.horizontalScreenEdgePadding = 10.0,
+    this.animationCurve = Curves.easeOut,
+  });
+
+  /// The curve used for the in / out animations.
+  final Curve animationCurve;
+
+  /// A drag resistance on the downward Y position of the lens.
+  final double dragResistance;
+
+  /// The difference in Y between the gesture position and the caret center
+  /// so that the magnifier hides itself.
+  final double hideBelowThreshold;
+
+  /// The padding on either edge of the screen that any part of the magnifier
+  /// cannot exist past.
+  ///
+  /// This includes the entire magnifier, not just the center.
+  ///
+  /// If the screen has width w, then the magnifier is bound to
+  /// `_kHorizontalScreenEdgePadding, w - _kHorizontalScreenEdgePadding`.
+  final double horizontalScreenEdgePadding;
+
+  /// The duration that the magnifier drags behind its final position.
+  static const Duration _kDragAnimationDuration = Duration(milliseconds: 45);
+
+  /// This magnifier's controller.
+  ///
+  /// The [CupertinoTextMagnifier] requires a [MagnifierController]
+  /// in order to show / hide itself without removing itself from the
+  /// overlay.
+  final MagnifierController controller;
+
+  /// [CupertinoTextMagnifier] will determine its own positioning
+  /// based on the [MagnifierOverlayInfoBearer] of this notifier.
+  final ValueNotifier<MagnifierOverlayInfoBearer>
+      magnifierOverlayInfoBearer;
+
+  @override
+  State<CupertinoTextMagnifier> createState() =>
+      _CupertinoTextMagnifierState();
+}
+
+class _CupertinoTextMagnifierState extends State<CupertinoTextMagnifier>
+    with SingleTickerProviderStateMixin {
+  // Initalize to dummy values for the event that the inital call to
+  // _determineMagnifierPositionAndFocalPoint calls hide, and thus does not
+  // set these values.
+  Offset _currentAdjustedMagnifierPosition = Offset.zero;
+  double _verticalFocalPointAdjustment = 0;
+  late AnimationController _ioAnimationController;
+  late Animation<double> _ioAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    _ioAnimationController = AnimationController(
+      value: 0,
+      vsync: this,
+      duration: CupertinoMagnifier._kInOutAnimationDuration,
+    )..addListener(() => setState(() {}));
+
+    widget.controller.animationController = _ioAnimationController;
+    widget.magnifierOverlayInfoBearer
+        .addListener(_determineMagnifierPositionAndFocalPoint);
+
+    _ioAnimation = Tween<double>(
+      begin: 0.0,
+      end: 1.0,
+    ).animate(CurvedAnimation(
+        parent: _ioAnimationController, curve: widget.animationCurve));
+  }
+
+  @override
+  void dispose() {
+    widget.controller.animationController = null;
+    widget.magnifierOverlayInfoBearer
+        .removeListener(_determineMagnifierPositionAndFocalPoint);
+    super.dispose();
+  }
+
+  @override
+  void didChangeDependencies() {
```

Comment:
Do you also need to add and remove the listener on magnifierOverlayInfoBearer in this method if it has changed?

Reply:
like this?

```dart 
  @override
  void didUpdateWidget(CupertinoTextMagnifier oldWidget) {
    if (oldWidget.magnifierOverlayInfoBearer != widget.magnifierOverlayInfoBearer) {
      oldWidget.magnifierOverlayInfoBearer.removeListener(_determineMagnifierPositionAndFocalPoint);
      widget.magnifierOverlayInfoBearer.removeListener(_determineMagnifierPositionAndFocalPoint);
    }
    super.didUpdateWidget(oldWidget);
  }
```

Reply:
added to both cupertino and material